enhancements/storage/csi-driver-install.md

```diff
@@ -0,0 +1,181 @@
+
+### Phase-1 Goals
+
+* Provide a way for AWS and GCE CSI driver installation . Users could either optionally install it or CSI driver could be installed
+along with in-tree driver and users could use both.
+* Install CSI provided storageclass along with in-tree StorageClass.
+
+### Phase-2 Goals
+
+* Support installation of GCE, Cinder, vSphere, AzureDisk CSI drivers.
+* Provide a way for users to enable in-tree to CSI migration. It basically replaces whole in-tree volume plugin with a less tested CSI implementation. While it's not a strict requirement, we'd like to have the migration optional for at least one release.
+* Provide a way for users to disable in-tree to CSI migration.
+
+### Phase-3 Goals
+
+* Enable CSI drivers as default drivers and CSI provided storageclass as default storageclass.
+* Optionally allow users to configure CSI driver install.
+
+## Non-Goals
+
+* This KEP does not attempt to design installation of third-party or partner CSI drivers.
+* Design of CSI migration enablement.
+  * A new operator will be needed to enable alpha features on API server, controller-manager and kubelets in the right order and with proper node draining before switching the gates.
+  * Right now it's implementation detail, this enhancement focuses on CSI driver installation. Proper enhancement will be created for CSI migration.
+
+## Proposal
+
+We are currently considering using OLM for installation of CSI driver.
+
+### Installation via OLM
+
+We propose that - we provide each driver mentioned above as a separate operator which could be subscribed and installed via OLM UI. Each driver's operator
+is responsible for its installation and release. The operator is responsible for creating storageclass that the driver provides.
+
+The configuration of CSI driver can be done via OLM UI if required and CSI driver can access cloudprovider credentials from Openshift provided sources.The CR that is responsible for driver configuration can be installed by default by the operator itself.
+
+Installation via OLM however means that, when we want to enable these CSI drivers as default drivers, they must be installed by default in Openshift installs.
+We further propose that - Cluster Storage Operator(https://github.com/openshift/cluster-storage-operator) could create subscriptions for these driver operators when drivers have to be installed by default.
+
+Expected workflow as optional driver (using EBS as an example):
+1. User finds EBS CSI driver in OLM and installs it.
+2. EBS CSI driver is installed and it creates relevant storageclass that user can use to provision CSI EBS volumes.
+
+Expected workflow as default driver:
+1. CVO installs cluster-storage-operator.
+2. cluster-storage-operator detects cloudprovider on which cluster is running(lets say EBS).
+3. cluster-storage-operator creates a subscription for EBS CSI driver using redhat operator source.
```

Comment:
IMO a separate workflow for *upgrade from opt-in to default CSI driver", when user installs an optional CSI driver operator in 4.5 and updates the cluster to version that has the driver installed by default.

Important step I'd like to emphasize that cluster-storage-operator must find existing subscription, potentially in a non-standard namespace, and "adopt it". It must ensure that the subscription always exists, even in that non-standard namespace. Or is it possible to "move" the operator to a different namespace?

Reply:
"moving" the operator requires some investigation, because I would imagine there would be a CR created by the operator install and that has to be deleted and re-created. The deletion of CR could trigger deletion of deployments/daemonsets managed by the operator, which could be problematic during upgrade.

Reply:
[just brainstorming here]
There is no CR moving if the CR is not namespaced and has predefined name. We could have a webhook that denies creation of any CR that does not have the predefined name.

Moving of the CSI driver e.g. between different namespaces is another story though. We could make the operator installable in any namespace, but it could install the driver into a hardcoded one (say `openshift-aws-ebs-csi-driver`), so we always know where it is. "Adoption" would just reinstall the operator into a defined namespace, while the driver itself could be untouched.

So it seems it could be possible to adopt the operator && move it to 'our' namespace. There are some details to iron out (operator re-installation probably removes CRD and thus the CR) and overall it looks quite ugly. Is it possible to make it nicer?

Reply:
IIRC most CRs that allow operator configuration are namespaced in openshift. We could ask for some guidance on @deads2k about this.

The other detail worth keeping in mind about installing the driver in a hardcoded namespace is, the RBAC policies specified in CSV are only applied to the namespace in which operator is installed and hence operator must apply RBAC policies required by the driver itself in different namespace. This is not very atypical but worth keeping in mind.

Reply:
For example - permission to list PVCs inside operator has to be a cluster-scoped permission rather than namespaced so as operator can hand out this permission to any arbitrary namespace in which driver is installed.

Reply:
I moved this thing to entire new section and updated the flow.

Reply:
Some (all?) `cluster-*` operators managed by CVO (e.g. `cluster-csi-snapshot-controller-operator`) have non-namespaced CRs and they watch only one CR named `cluster`.